src/refdb_reftable.c

```diff
@@ -0,0 +1,514 @@
+{
+	git_reference *ref;
+	git_oid oid, peeled;
+	int error;
+
+	if (record->target) {
+		ref = git_reference__alloc_symbolic(record->ref_name, record->target);
+	} else {
+		if ((error = git_oid_fromraw(&oid, record->value)) < 0 ||
+		    (error = git_oid_fromraw(&oid, record->target_value)) < 0)
+			goto out;
+		ref = git_reference__alloc(record->ref_name, &oid, &peeled);
+	}
+	GIT_ERROR_CHECK_ALLOC(ref);
+
+	*out = ref;
+out:
+	return error;
+}
+
+static int reftable_log_fill(struct log_record *out,
+			     const git_signature *who,
+			     const git_oid *old_id,
+			     const git_oid *new_id,
+			     const char *reference,
+			     const char *message)
+{
+	memset(out, 0, sizeof(*out));
+	if (who) {
+		out->name = who->name;
+		out->email = who->email;
+		out->time = who->when.time;
+		out->tz_offset = who->when.offset;
+	}
+	if (old_id)
+		out->old_hash = (unsigned char *) old_id->id;
+	if (new_id)
+		out->new_hash = (unsigned char *) new_id->id;
+	out->ref_name = (char *) reference;
+	out->message = (char *) message;
+	return 0;
+}
+
+static int reftable_writer(struct writer *wr, void *_arg)
+{
+	reftable_writer_arg *arg = (reftable_writer_arg *) _arg;
+	writer_set_limits(wr, arg->index, arg->index + arg->nrefs + arg->nlogs);
+	if (writer_add_refs(wr, arg->refs, arg->nrefs) < 0)
+		return -1;
+	if (writer_add_logs(wr, arg->logs, arg->nlogs) < 0)
+		return -1;
+	return 0;
+}
+
+static int reftable_write_records(git_refdb_backend *_backend,
+				  struct ref_record *refs, int nrefs,
+				  struct log_record *logs, int nlogs)
+{
+	reftable_backend *backend = GIT_CONTAINER_OF(_backend, reftable_backend, parent);
+	reftable_writer_arg arg = {
+		refs, logs, nrefs, nlogs, stack_next_update_index(backend->stack)
+	};
+
+	if (stack_add(backend->stack, reftable_writer, &arg) < 0)
+		return -1;
+
+	return 0;
+}
+
+static int reftable_exists(
+	int *exists,
+	git_refdb_backend *_backend,
+	const char *ref_name)
+{
+	reftable_backend *backend = GIT_CONTAINER_OF(_backend, reftable_backend, parent);
+	struct ref_record ref = { 0 };
+	int error;
+
+	if ((error = stack_read_ref(backend->stack, ref_name, &ref)) < 0)
+		return error;
+	*exists = (error == 0);
+
+	ref_record_clear(&ref);
+	return 0;
+}
+
+static int reftable_lookup(
+	git_reference **out,
+	git_refdb_backend *_backend,
+	const char *ref_name)
+{
+	reftable_backend *backend = GIT_CONTAINER_OF(_backend, reftable_backend, parent);
+	struct ref_record ref;
+	int error;
+
+	if ((error = stack_read_ref(backend->stack, ref_name, &ref)) < 0)
+		return error;
+	if (error > 0)
+		return GIT_ENOTFOUND;
+
+	return reftable_reference_from_record(out, &ref);
+}
+
+static int reftable_iterator_next(git_reference **out, git_reference_iterator *_it)
+{
+	reftable_iterator *it = GIT_CONTAINER_OF(_it, reftable_iterator, parent);
+	struct ref_record ref;
+	int error;
+
+	while (1) {
+		if ((error = iterator_next_ref(it->iter, &ref)) < 0)
+			return error;
```

Comment:
shouldn't you translate error spaces somewhere?

Reply:
Indeed, I've got that on my todo list.